lib/main.js

```diff
@@ -1,156 +1,87 @@
-      idleCallbacks.delete(createWorkerCallback);
+  startServerProcess(projectPath) {
+    return super.spawnChildNode(
+      [require.resolve('@keplersj/tslint-server/out/server')],
+      { cwd: projectPath, stdio: [null, null, null, 'ipc'] },
+    );
+  }
+
+  preInitialization(connection) {
+    connection._onNotification({ method: 'tslint/status' }, (params) => { // eslint-disable-line no-unused-vars
+      // This notification is used to set the status indicator in VSCode.
+
+      // const status = params.state;
+      // switch (status) {
+      //   case 1: // ok
+      //     atom.notifications.addSuccess("TSLint is OK");
+      //     break;
+      //   case 2: // warning
+      //     atom.notifications.addWarning("TSLint is not OK");
+      //     break;
+      //   case 3: // error
+      //     atom.notifications.addError("TSLint is not OK");
+      //     break;
+      //   default:
+      //     console.error("TSLint status changed in an unknown way!")
+      // }
     });
-    idleCallbacks.add(createWorkerCallback);
-  },
-
-  deactivate() {
-    idleCallbacks.forEach(callbackID => window.cancelIdleCallback(callbackID));
-    idleCallbacks.clear();
-    this.subscriptions.dispose();
-
-    workerHelper.terminateWorker();
-    this.worker = null;
-  },
-
-  provideLinter() {
-    return {
-      name: 'TSLint',
-      grammarScopes,
-      scope: 'file',
-      lintOnFly: true,
-      lint: async (textEditor) => {
-        if (this.ignoreTypings && textEditor.getPath().toLowerCase().endsWith('.d.ts')) {
-          return [];
-        }
-
-        if (!this.worker) {
-          await waitOnIdle();
-        }
 
-        workerHelper.startWorker(this.worker, config);
-
-        const text = textEditor.getText();
-        const results = await workerHelper.requestJob('lint', textEditor);
-
-        if (textEditor.getText() !== text) {
-          // Text has been modified since the lint was triggered, tell linter not to update
-          return null;
-        }
+    connection._onRequest({ method: 'tslint/noLibrary' }, () => {
+      atom.notifications.addError(
+        "TSLint can't find TSLint Library in Project.",
+      );
+    });
 
-        return results;
-      },
-    };
-  },
-};
+    connection._onRequest({ method: 'workspace/configuration' }, () => ({
+      enable: true,
+      jsEnable: atom.config.get('linter-tslint.javascriptSupport'),
+      rulesDirectory: atom.config.get('linter-tslint.rulesDirectory'),
+      configFile: path.join(this.projectPath, 'tslint.json'),
+      ignoreDefinitionFiles: atom.config.get('linter-tslint.ignoreTypings'),
+      exclude: [''],
+      validateWithDefaultConfig: !atom.config.get('linter.tslint.useLocalTslint'),
+      nodePath: process.execPath,
+      run: 'onType',
+      alwaysShowRuleFailuresAsWarnings: false,
+      autoFixOnSave: atom.config.get('linter-tslint.fixOnSave'),
+      trace: atom.config.get('core.debugLSP') ? { server: 'verbose' } : {},
+    }));
```

Comment:
Nice use of onRequest to implement workspace configuration!  Given that project specific settings in atom aren't a thing I wonder if we can do something to facilitate something similar with less effort just for our providers in lsp.

Reply:
should also observe these config values and send a 'workspace/didChangeConfiguration' when that happens.

Reply:
Awesome idea. Done!